admin/test/test_merge_pr.py

```diff
@@ -0,0 +1,212 @@
+        self.assertIn("Merge a branch when all the tests pass.", output)
+
+
+class URLTests(TestCase):
+    """
+    Test for URL manipulation.
+    """
+
+    def test_url_path(self):
+        """
+        Correct path from a full URL.
+        """
+        path = '/ClusterHQ/flocker/pull/1717'
+        self.assertEqual(path, merge_pr.url_path('https://github.com' + path))
+
+    def test_url_path_no_hostname(self):
+        """
+        Correct path from a URL path.
+        """
+        path = '/ClusterHQ/flocker/pull/1717'
+        self.assertEqual(path, merge_pr.url_path(path))
+
+    def test_url_path_parts(self):
+        """
+        Correct segments from a full URL.
+        """
+        path_parts = ['ClusterHQ', 'flocker', 'pull', '1717']
+        self.assertEqual(
+            [''] + path_parts,
+            merge_pr.url_path_parts(
+                'https://github.com/' + '/'.join(path_parts)))
+
+    def test_url_path_parts_no_hostname(self):
+        """
+        Correct segments from a URL path.
+        """
+        path_parts = ['ClusterHQ', 'flocker', 'pull', '1717']
+        self.assertEqual(
+            [''] + path_parts,
+            merge_pr.url_path_parts('/' + '/'.join(path_parts)))
+
+    def test_pr_api_url(self):
+        """
+        Correct API URL for a full URL.
+        """
+        self.assertEqual(
+            'https://api.github.com/repos/ClusterHQ/flocker/pulls/1717',
+            merge_pr.pr_api_url_from_web_url(
+                'https://github.com/ClusterHQ/flocker/pull/1717'))
+
+
+# The max here corresponds to the latest date that datetime supports
+datetimes = integers(max_value=253402300799).map(datetime.fromtimestamp)
+"""Strategy for generating `datetime` objects."""
+
+
+commit_statuses = fixed_dictionaries(
+    {'updated_at': datetimes,
+     'state': text(),
+     'context': text(),
+     'target_url': text(),
+     })
+"""Strategy for generating GitHub commit status dicts."""
+
+
+jenkins_results = sampled_from(merge_pr.JenkinsResults.iterconstants())
+"""Strategy for generating JenkinsResults values."""
+
+
+class StatusesTests(TestCase):
+    """
+    Tests for interpretation of commit statuses.
+
+    https://developer.github.com/v3/repos/statuses/
+    """
+
+    @given(commit_statuses)
+    def test_final_status_one(self, status):
+        self.assertEqual(status, merge_pr.final_status([status]))
+
+    @given(commit_statuses, commit_statuses)
+    def test_final_status_many(self, status1, status2):
+        target = status1
+        if status2['updated_at'] > status1['updated_at']:
+            target = status2
+        self.assertEqual(target, merge_pr.final_status([status2, status1]))
+
+    @given(commit_statuses)
+    def test_not_success(self, status):
+        assume(status['state'] != u"success")
+        self.assertEqual(True, merge_pr.not_success(status))
+
+    @given(commit_statuses)
+    def test_not_success_success(self, status):
+        status['state'] = u"success"
```

Comment:
:question: I think it's slightly better to use a different strategy:
```
fixed_dictionaries(
     {'updated_at': datetimes,
      'state': just('success'),
      'context': text(),
      'target_url': text(),
      })
```
Not required for merge, but note that as the code is currently, this test might be run many times where hypothesis only varies the state key, which will be sort of lost testing.

Might consider turning commit_statuses into a function that lets you override the strategies:
```
def commit_statuses(**kwargs):
  return fixed_dictionaries(dict({
     'updated_at': datetimes,
      'state': text(),
      'context': text(),
      'target_url': text(),
      }, **kwargs))
```